x-pack/legacy/plugins/siem/server/utils/build_query/calculate_timeseries_interval.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import moment from 'moment';
+import { get } from 'lodash/fp';
+const d = moment.duration;
+
+const roundingRules = [
+  [d(500, 'ms'), d(100, 'ms')],
+  [d(5, 'second'), d(1, 'second')],
+  [d(7.5, 'second'), d(5, 'second')],
+  [d(15, 'second'), d(10, 'second')],
+  [d(45, 'second'), d(30, 'second')],
+  [d(3, 'minute'), d(1, 'minute')],
+  [d(9, 'minute'), d(5, 'minute')],
+  [d(20, 'minute'), d(10, 'minute')],
+  [d(45, 'minute'), d(30, 'minute')],
+  [d(2, 'hour'), d(1, 'hour')],
+  [d(6, 'hour'), d(3, 'hour')],
+  [d(24, 'hour'), d(12, 'hour')],
+  [d(1, 'week'), d(1, 'd')],
+  [d(3, 'week'), d(1, 'week')],
+  [d(1, 'year'), d(1, 'month')],
+  [Infinity, d(1, 'year')],
+];
+
+const revRoundingRules = roundingRules.slice(0).reverse();
+
+const find = (
+  rules: Array<Array<number | moment.Duration>>,
+  check: (
+    bound: number | moment.Duration,
+    interval: number | moment.Duration,
+    target: number
+  ) => number | moment.Duration | undefined,
+  last?: boolean
+): ((buckets: number, duration: number | moment.Duration) => moment.Duration | undefined) => {
+  function pick(buckets: number, duration: number | moment.Duration): number | moment.Duration {
+    const target =
+      typeof duration === 'number' ? duration / buckets : duration.asMilliseconds() / buckets;
+    let lastResp = null;
+
+    for (let i = 0; i < rules.length; i++) {
+      const rule = rules[i];
+      const resp = check(rule[0], rule[1], target);
+
+      if (resp == null) {
+        if (last) {
+          if (lastResp) return lastResp;
+          break;
+        }
+      }
+
+      if (!last && resp) return resp;
+      lastResp = resp;
+    }
+
+    // fallback to just a number of milliseconds, ensure ms is >= 1
+    const ms = Math.max(Math.floor(target), 1);
+    return moment.duration(ms, 'ms');
+  }
+
+  return (buckets, duration) => {
+    const interval = pick(buckets, duration);
+    const intervalData = get('_data', interval);
+    if (intervalData) return moment.duration(intervalData);
+  };
+};
+
+export const calculateAuto = {
+  near: find(
+    revRoundingRules,
+    function near(bound, interval, target) {
+      if (bound > target) return interval;
+    },
+    true
+  ),
+
+  lessThan: find(revRoundingRules, function lessThan(_bound, interval, target) {
+    if (interval < target) return interval;
+  }),
+
+  atLeast: find(revRoundingRules, function atLeast(_bound, interval, target) {
+    if (interval <= target) return interval;
+  }),
+};
+
+export function calculateTimeseriesInterval(
```

Comment:
I would at the least for this hard fork remove the function in favor of ES2015/ES6 function styles we have been using as we might eventually turn on a linter rule to forbid the older named functions due to their problems with the `self` and `this` compared to the fat arrow functions.